### PR TITLE
[16.0][MIG] stock_barcodes: Migration to 16.0

### PR DESCRIPTION
**Pull request by chienandalu**
Followup to #587

- Drop scan read logs feature, which use was deprecated already in the
  previous versions.
- Pass bootstrap migration scripts.
- Fix views issues.
- Other code fixes.

cc @Tecnativa

@FranzPoize I've been doing the migration with @sergio-teruel, who's the main developer of the module in the spanish OCA days. We started from your migration head.

At this point the module is completely functional, although we're still tweaking the views to fit the previous state.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
This doesn't contain #604

**Comment by chienandalu** [comment]
> This doesn't contain #604

Fixed

**Comment by FranzPoize** [comment]
Hello ! I pushed some stuff just after you created this PR I think. In it there was a lot of work to replace the `always_reload` feature of old odoo version.

**Comment by chienandalu** [comment]
Hi, @FranzPoize !

Do you mean this new stuff or something else?: https://github.com/OCA/stock-logistics-barcode/pull/587/commits/38288280b173ffc2487758a85edad3bb74959442#diff-f69121127ec9ebdc28296fa3a7f3760d71865de0267308101bf6d467fef4769cR106-R121

**Comment by FranzPoize** [comment]
No I mean this stuff https://github.com/OCA/stock-logistics-barcode/commit/38288280b173ffc2487758a85edad3bb74959442#diff-00871f58b13dd92c5ff7599bc210323e9a6e64bdbf58abd5a3c19c83428d5dbb

What you linked to is some other features we added:
* Allowing scanning product on at a time in new picking (not use the total quant by default)
* Forbidding picking validation if their is a move_line with the same location_id and location_dest_id
* Offer an option to not check demand during picking (this solve the issue where if you reload after a picking creation you will not be able to pick more product without using the force action done)
* Update all destination of all move_line when scanning a location (This allows destination to be scanned last because otherwise it does not work)

**Comment by chienandalu** [comment]
> there was a lot of work to replace the always_reload feature of old odoo version.
> I mean this stuff https://github.com/OCA/stock-logistics-barcode/commit/38288280b173ffc2487758a85edad3bb74959442#diff-00871f58b13dd92c5ff7599bc210323e9a6e64bdbf58abd5a3c19c83428d5dbb

Can you elaborate on why this is needed? What's the issue you're trying to solve?

> other features we added:
> - Allowing scanning product on at a time in new picking (not use the total quant by default)
> - Forbidding picking validation if their is a move_line with the same location_id and location_dest_id
> - Offer an option to not check demand during picking (this solve the issue where if you reload after a picking creation you will not be able to pick more product without using the force action done)
> - Update all destination of all move_line when scanning a location (This allows destination to be scanned last because otherwise it does not work)

I commented these with @sergio-teruel last week and some of them could be accomplished with the current config (I'm sure he can elaborate in detail :)). In any case, is better to limit the scope the PR to the module migration stuff and do any feature proposal in a later one

**Comment by FranzPoize** [comment]
You don't have to take those features if you don't want to. But you should take the `always_reload` feature imho because otherwise there is no update of many2one field unless there is a button pressed.

**Comment by FranzPoize** [comment]
In 14.0 there was this https://github.com/OCA/stock-logistics-barcode/blob/fe5c849fc9a6ec190eed2672086df22553c78d64/stock_barcodes/wizard/stock_barcodes_read_picking_views.xml#L160 This does not work anymore the code here https://github.com/OCA/stock-logistics-barcode/commit/38288280b173ffc2487758a85edad3bb74959442#diff-00871f58b13dd92c5ff7599bc210323e9a6e64bdbf58abd5a3c19c83428d5dbb replaces this always_reload feature.

And I said many2ones in my previous comment I made a mistake. This allows one2many fields and many2many field to update their data when an onchange event is triggered.

**Comment by chienandalu** [comment]
Hi @FranzPoize let me check it and I tell you

**Comment by lk-eska** [comment]
Is there any progress on this PR?

**Comment by sergio-teruel** [comment]
Hi @ik-eska two commits 4 days ago... Any progress?? Yes...

**Comment by lk-eska** [comment]
Sorry @sergio-teruel I haven't noticed them.

**Comment by sergio-teruel** [comment]
jaja... don't worry... :smile:

**Comment by sergio-teruel** [comment]
Cherry pick from 15.0 ef65da1c6e432e37dd045a77314e79c74a60fe9b

**Comment by pedrobaeza** [comment]
@chienandalu @sergio-teruel please fix the CI and finish this for merging it.

**Comment by josepmy** [comment]
Se puede ayudar en algo para avanzar el merge?

**Comment by pedrobaeza** [comment]
@sergio-teruel please put this green to be merged.

**Comment by sergio-teruel** [comment]
@pedrobaeza Tests ok

**Comment by pedrobaeza** [comment]
@carlosdauden should I merge or should I wait for your last patches? Can you give your approval?

**Comment by abhinavvaidya-tsd** [comment]
is this PR stuck?

**Comment by carlosdauden** [comment]
We plan to finish the work by the end of this week.

**Review by houzefa-abba — APPROVED** [review]
Thanks for this work, I tested on runboat / in local env:

* Scan interface works fine with my Inateck BCST-23 without any specific config
* "Picking IN" mode with the "Create lots if not match" setting
* "Picking OUT" mode
* "Show Detailed Operations" setting
* Disable "Fill Fields From Lot" setting to scan the same product several times if multiple quantities are expected
* Play with combos of these settings to scan multiple serial numbers of the same product within 1 picking
* Scan **more** quantities than expected, get "quantities higher than necessary" message, click on red ✔️ to force adding it
* Scan **less** quantities than expected, click "Validate" early on ➔ Partial transfer gets validated & second one is created with remaining quantities

**Comment by houzefa-abba** [comment]
Hi I have this issue in inventory mode:
* Give a product a barcode + tracking by serial number
* Go to Barcodes > Inventory
* Scan the product barcode added above

➔ Error:
```
  File "/mnt/data/odoo-addons-dir/stock_barcodes/wizard/stock_barcodes_read.py", line 142, in _compute_qty_available
    done_move_lines = self.move_line_ids.filtered(
AttributeError: 'wiz.stock.barcodes.read.inventory' object has no attribute 'move_line_ids'
```

In 15.0, this displays an "Expected serial number scan" popup, but does not crash.

Also note:
* No error when the product has no tracking by s/n
* No error when scanning a serial number instead

**Comment by sergio-teruel** [comment]
Hi @houzefa-abba Fixed in last commit (4978de68945d25d9f6dcf07beb269b11c666817e)

Thanks

**Comment by alexis-via** [comment]
Thanks for your great work on the migration of this great module.

I have a usability remark on inventory: you added a new boolean option "display_read_quant" on the wizard wiz.stock.barcodes.read.inventory. When this option is not checked, the user doesn't see the result of the scans, which is error-prone.
So I suggest 3 solutions:
1) modify the definition of this field to set default=True
2) move this option to stock.barcodes.option.group, so that we can enable this option for all inventories
3) add an option on stock.barcodes.option.group that will act as a default value for the field "display_read_quant" of the wizard (combination of solution 1 and 2)
I don't have a strong opinion myself about the best option. As of today, my opinion is : it should not be an option, because I think it's too risky to scan without visibility on the result !
If you want, I'm ready to make a PR to your branch to implement the preferred solution.

File: stock_barcodes/models/stock_quant.py
Comment (alexis-via) [inline stock_barcodes/models/stock_quant.py:30]: When reading this code, I see that, on line 30, in the loop on quant._fields that are in _get_fields_to_edit(), it will write:
```
wiz_barcode.product_qty
```
then, on line 31, it will write:
```
wiz_barcode.product_qty = quant.inventory_quantity
```
Is it normal ? Shouldn't we remove "product_qty" from _get_fields_to_edit ?

**Comment by sergio-teruel** [comment]
Hi @alexis-via thanks for your comments

I have added display read items in the option group to apply as a default value.
Also fix the view and your suggestion in stock quant

**Comment by alexis-via** [comment]
@sergio-teruel I made some tests with your last commit: it works very well, thanks a lot !
Thanks for fixing the view with the display of the quants!

**Review by alexis-via — APPROVED** [review]

**Review by carlosdauden — APPROVED** [review]

**Pull request merged**
